### ISSUE TITLE
Adding a Refresh() to IServiceProvider [Dependency Injection]

### ISSUE THREAD

**Issue by FrederickBrier** [aspnet/Home#2766]
After adding most of my services in ConfigureServices(), One of the services is instantiated using app.ApplicationServices.GetServices<> at the beginning of Configure(). This utilizes the ServiceProvider instance. This occurs at the beginning of the method, and the service adds some more services to ServiceCollection which was preserved in the original ConfigureServices() call. The problem is that these new services are not available via the ServiceProvider instance. Creating another ServiceProvider instance from the ServiceCollection would lose previously modified and needed singletons.

 I downloaded the DependencyInjection, Metadata, and Hosting github repos. Digging into the ServiceProvider, ultimately revealed a reference to the ServiceCollection is preserved in the internal property and class CallSiteFactory inside the internal base class ServiceProviderEngine. The private method Populate() initializes the _descriptorLookup data member. It would be nice for ServiceProvider to have a Refresh() method that would call down into the CallSiteFactory and blow away and call Populate() to recreate the _descriptorLookup data member. If I am reading the code correctly, this would not affect the singleton instances, and the recreated cacheKey(s) would be the same.

If this would work and is acceptable, I would be glad to fork and try it out.

**Comment by Eilon** [comment]
@pakrym / @davidfowl - thoughts on this?

**Comment by pakrym** [comment]
Not supported by design. There are some containers that support this scenario (NInject?) but not sure if any of them supports `Microsoft.Extensions.DependencyInject.Abstrations`.

**Comment by FrederickBrier** [comment]
Unfortunate. My PluginManager is registered with the ServiceCollection, but after it is instantiated, it needs to register additional types loaded from each plugin. No more registrations occur after that, but it is a two phase process.

**Comment by muratg** [comment]
This is by design so closing.  @FrederickBrier let me know if you'd like to talk about alternative ways to implement that.

**Comment by FrederickBrier** [comment]
pakrym suggested looking at NInject. I found [someone](https://github.com/msdevno/GeekBeer/blob/master/Web/Ninject/ServiceProviderNinjectModule.cs) who had created a way for NInject to consume the ServiceDescriptors. I then made some small adjustments. Being able to register the classes in the plugin and have NInject populate the plugin's main class constructor with all the objects previously registered removed the use of reflection. The code was cleaner and more readable. But thank you for the offer. Your team has been very responsive.
